### DISCUSSION TITLE
Simulate a Tube Robot

### DISCUSSION THREAD

**Discussion by LMI-pewpew** [GazzolaLab/PyElastica#60] [Q&A] [answered]
Hi there,

Im searching for a way to simulate a concentric tube robot and I stumbled upon PyElastica.

Can I simulate tubes using this software?  Its not clear to me that i can after going through your website.

This may be a very simple question, but I wanted to be sure before I dive deeper into PyElastica.

Thanks!

**Comment by skim0119** [comment, marked as answer]
@LMI-pewpew Hi, 

Yes, you can simulate tubes. I'm also working on a softrobot project that uses a silicon tube, and I'm using PyElastica as a simulation tool. I use `CosseratRod` in the package and manually modify the `area`, `bend_matrix`, and `shear_matrix`. 

In my experience, the simulation gives accurate kinematics as long as there is no buckling. Also, I wouldn't expect the change in radius to be accurate, since the original `CosseratRod` assumes a circular (not hollow) cross-section, but I found this is not significant for a mild deformation.

**Reply by LMI-pewpew** [reply]
Thanks @skim0119 for the very useful information,

What do you consider mild deformation?  I am working with sub millimeter diameter tubes where the bend radii of the tubes range from 20 to 0 mm.

**Reply by skim0119** [reply]
I mean __mild deformation__ as there is no buckling or any bifurcation (like collapsing). I don't have any quantitative study at this point, and it would vary from case to case.

I'm using a tube of a diameter 1 cm (20-30cm length) and bend radii of about 5 cm (I haven't tested further). The simulation result with PyElastica _roughly_ match the experiment. I'm saying `rougly` because my instrument measuring the position has some error on its own (The instrument has 2-4% error, while the simulation matched within 5%; equivalent to maximum 0-0.5mm position error). Overall configuration (geometry and morphology) and kinematic matches. In my case, I activate the tube using the internal pressure (hydraulic and pneumatic), so it would have more resistance to the buckling behavior.

I would say the best approach is to build a quick simulation using built-in modules and see if the result approximately matches within the workspace of your robot. If the result is promising, refine the `boundary conditions` and `forcing` to match the experiment, and measure the precision.

**Reply by LMI-pewpew** [reply]
Thanks for all the info, much appreciated!!
